### ISSUE TITLE
Alex Dowad

### ISSUE THREAD

**Issue by krainboltgreene** [plexus/rubydataspec#5]
This guy is an absolute powerhouse in terms of activity and knowledge. I bet he'd be interested in this.

github: https://github.com/alexdowad

**Comment by plexus** [comment]
ping @alexdowad

**Comment by alexdowad** [comment]
Hi, just saw this. So the purpose of this project is to ensure compatibility is maintained between multiple persistent data structure libraries? Do you have buy-in from the maintainers of all the alternative libraries?

**Comment by plexus** [comment]
Yes, I've worked with @Who828, the implementer of Persitent::Vector, on this. The maintainer of JRuby, @headius is the one who did Clojr, and he's also interested. The JRuby folks in particular are interested in persistent data structures, since they have actual concurrency, and they're not afraid of "standards" or "interfaces" the way some Ruby folks are.

There's not a whole lot of traction yet. I was hoping to step by step agree on the "spec", then make implementations conform, but I've since realized that the only way to get meaningful feedback is by making pull requests for the implementations, for which so far I haven't had the time.

**Comment by alexdowad** [comment]
Making different implementations compatible is a good thing, as long as the
design goals of each library are close enough to each other that they are
not "contorted" (i.e. forced to adopt an API which feels unnatural for that
particular library).

Making this effort will only be meaningful if the spec is constraining
enough to cover most common uses of the libraries in question.

I wonder: would not be be better to produce one library which is so good
that there would be no question of wanting a different one? After all, one
good choice is better than 10 mediocre choices. It can have "drop-in
replacement" C/Java implementations for speed.

Is there a good argument to have more than 1 library to do what is
essentially the same thing?

**Comment by plexus** [comment]
You already point out the reason, people use persistent data structures for performance. You could have a single project that implement both a C version for MRI and a Java version for JRuby, but as it stands a few people are only targeting their platform of choice. I think it's good to have a few different projects, so we can benchmark different use cases and figure out the strengths and weaknesses of each.

Taking a lead from Clojure, it is possible to categorize data structures in a few categories that imply a certain minimal interface, sequential (list, vector), accessible by index/key (vector, hash), etc. Specific implementations can add more operations, but e.g. when I get something back from a library I can use it simply by knowing that it behaves like a sequential data structure, whatever the implementation (which might change in the next version of the library).

You can mimic Ruby's core data structures, but not always, e.g. you can never use the foo[x]=bar syntax for a persistent hash, so what to use instead? put, set, conj? I think it's useful to pick one and encourage everyone to use it.

**Comment by alexdowad** [comment]
Well, you have some good points. I will not be the person to come up with the spec you are describing, but once you have a draft, I would like to review it. Why not write up a document and send it around to everyone you know who may have useful feedback? It doesn't need to be a legal treatise -- just a brief description of each API method and some short code samples.

Here's some feedback for you straight away: I see your "simplest thing that could work" `Vector` implementation offers a `#prepend` operation. I don't see any good way to make `#prepend` efficient on the type of trie used for Clojure-style persistent vectors. Unless there is a way to make it efficient (at least it should run in logarithmic time), I would question including that operation in the API. Library users may assume that it's fast, then be unpleasantly surprised.
